### PR TITLE
docs(unit tests): Improve unit testing documentation

### PR DESCRIPTION
**Pull request by lucperkins**
The initial "new" vector.dev provided only thin documentation of Vector's unit testing feature. This PR provides much more thorough information.

The approach in this PR is a very "standard" Markdown-driven approach. At some point I would like to convert some of this into CUE-driven structured data but I haven't yet found a good way to do that. But I believe that the approach here provides a reasonable stopgap.

### CODE REVIEW

File: docs/content/en/docs/reference/configuration/tests.md [hunk 1]
```diff
@@ -5,14 +5,128 @@ weight: 5
...
 
-Vector enables you to unit test [transforms] in your processing topology
+Vector enables you to unit test [transforms] in your processing topology. The goal of unit tests in
+general is to improve the maintainability of configurations containing larger and more complex
+combinations of transforms.
 
-You can define unit tests in your Vector configuration file to cover a network of transforms within the topology. The intention of these tests is to improve the maintainability of configurations containing larger and more complex combinations of transforms.
+Unit tests in Vector work just like unit tests for standard software libraries:
+
+1. Provide a set of **inputs** to a transform (or a network of transforms)
```
Comment (spencergilbert) [inline docs/content/en/docs/reference/configuration/tests.md:14]: Network feels weird to me here as I'm not sure we use that to describe a set of transforms anywhere - do we want to reuse DAG/graph language here?

File: docs/content/en/docs/reference/configuration/tests.md [hunk 2]
```diff
@@ -5,14 +5,128 @@ weight: 5
...
+1. Provide expected **outputs** from the changes made by the transform(s)
+1. Receive directly actionable feedback on test failures
+
+If *any* of your unit tests fail, the Vector instance doesn't start up. This makes unit tests a
```
Comment (spencergilbert) [inline docs/content/en/docs/reference/configuration/tests.md:18]: This is another odd one to me as I don't think of any place were unit tests are run prior to startup - we have at least one place where `validate` is run though.

**Comment by lucperkins** [comment]
@spencergilbert Please note the draft status. This is not ready for review.

**Comment by jszwedko** [comment]
@lucperkins just wanted to see if this is still on your radar. Currently the unit test documentation is very sparse so I'm concerned about users discovering this feature. Up until now, it has been a very popular one.

File: docs/content/en/docs/reference/configuration/tests.md (hunk 2 above)
Comment (jszwedko) [reply docs/content/en/docs/reference/configuration/tests.md:18]: Agreed, unit tests don't run as part of Vector start-up.

File: docs/content/en/docs/reference/configuration/tests.md (hunk 1 above)
Comment (jszwedko) [reply docs/content/en/docs/reference/configuration/tests.md:14]: This appears to be language held over from the old doc, but I think just `set` would be fine here.

**Comment by spencergilbert** [comment]
> @lucperkins just wanted to see if this is still on your radar. Currently the unit test documentation is very sparse so I'm concerned about users discovering this feature. Up until now, it has been a very popular one.

Alternatively someone else can take over the PR and push it to completion

**Comment by aquarhead** [comment]
Awesome work! I've been thinking about contributing to unit test related docs too since it's kinda outdated, though I probably need to play with it a bit more myself before feeling comfortable to contribute, but let me know if you want me to take over?

**Comment by lucperkins** [comment]
Ended up with a merge catastrophe here due to lapse of time. Going to close and open a fresh PR.

**Comment by zsherman** [comment]
@lucperkins did a new PR come through?

**Pull request closed without merging**